src/ledger/device.cpp

```diff
@@ -0,0 +1,1679 @@
+  bool Device::reset() {
+    int offset;
+
+    lock_device();
+    reset_buffer();
+
+    this->buffer_send[0] = 0x00;
+    this->buffer_send[1] = INS_RESET;
+    this->buffer_send[2] = 0x00;
+    this->buffer_send[3] = 0x00;
+    this->buffer_send[4] = 0x00;
+    offset = 5;
+    //options
+    this->buffer_send[offset] = 0x00;
+
+    this->buffer_send[4] = offset-5;
+    this->length_send = offset;
+    this->exchange();
+    unlock_device();
+    return true;
+  }
+
+  /* ======================================================================= */
+  /*                             WALLET & ADDRESS                            */
+  /* ======================================================================= */
+
+  /* Application API */
+   bool Device::get_public_address(cryptonote::account_public_address &pubkey){
+    int offset;
+
+    lock_device();
+    reset_buffer();
+
+    this->buffer_send[0] = 0x00;
+    this->buffer_send[1] = INS_GET_KEY;
+    this->buffer_send[2] = 0x01;
+    this->buffer_send[3] = 0x00;
+    this->buffer_send[4] = 0x00;
+    offset = 5;
+    //options
+    this->buffer_send[offset] = 0x00;
+    offset += 1;
+
+    this->buffer_send[4] = offset-5;
+    this->length_send = offset;
+    this->exchange();
+
+    memmove(pubkey.m_view_public_key.data, this->buffer_recv, 32);
+    memmove(pubkey.m_spend_public_key.data, this->buffer_recv+32, 32);
+    unlock_device();
+    return true;
+  }
+
+  bool  Device::get_secret_keys(crypto::secret_key &viewkey , crypto::secret_key &spendkey) {
+    #ifdef DEBUGLEDGER
+    int offset;
+
+    lock_device();
+    reset_buffer();
+
+    this->buffer_send[0] = 0x00;
+    this->buffer_send[1] = INS_GET_KEY;
+    this->buffer_send[2] = 0x02;
+    this->buffer_send[3] = 0x00;
+    this->buffer_send[4] = 0x00;
+    offset = 5;
+    //options
+    this->buffer_send[offset] = 0x00;
+    offset += 1;
+
+    this->buffer_send[4] = offset-5;
+    this->length_send = offset;
+    this->exchange();
+
+    //encryptedkey or clearkey
+    memmove(viewkey.data,  this->buffer_recv+0,   32);
+    memmove(spendkey.data, this->buffer_recv+32, 32);
+    //clear key
+    memmove(ledger::viewkey.data,  this->buffer_recv+64, 32);
+    memmove(ledger::spendkey.data, this->buffer_recv+96, 32);
+    //spcialkey, normal conf handled in decrypt
+    memset(viewkey.data, 0x00, 32);
+    memset(spendkey.data, 0xFF, 32);
+    unlock_device();
+    #else
+    memset(viewkey.data, 0x00, 32);
+    memset(spendkey.data, 0xFF, 32);
+    #endif
+    return true;
+  }
+
+  bool  Device::get_chacha8_prekey(char  *prekey)  {
+    int offset;
+
+    lock_device();
+    reset_buffer();
+    this->buffer_send[0] = 0x00;
+    this->buffer_send[1] = INS_GET_CHACHA8_PREKEY;
+    this->buffer_send[2] = 0x00;
+    this->buffer_send[3] = 0x00;
+    this->buffer_send[4] = 0x00;
+    offset = 5;
+    //options
+    this->buffer_send[offset] = 0x00;
+    offset += 1;
+
+    this->buffer_send[4] = offset-5;
```

Comment:
This writes into offset 4 again, instead of 6, over-writing what was stored on line 523. Is that intentional?

Reply:
Yes It is. Code is today not optimized to help debug and update.

Command in buffer_send is 

        00(u8) INS(u8) P1(u8) P1(u8) LEN(8) PAYLOAD(LEN bytes)